kotsu/run.py

```diff
@@ -95,16 +94,26 @@ def run(
             model = model_spec.make()
             results, elapsed_secs = _run_validation_model(validation, model, run_params)
             results = _add_meta_data_to_results(results, elapsed_secs, validation_spec, model_spec)
-            results_list.append(results)
-
-    additional_results_df = pd.DataFrame.from_records(results_list)
-    results_df = pd.concat([results_df, additional_results_df], ignore_index=True)
-    results_df = results_df.drop_duplicates(subset=["validation_id", "model_id"], keep="last")
-    results_df = results_df.sort_values(by=["validation_id", "model_id"]).reset_index(drop=True)
-    store.write(
-        results_df, results_path, to_front_cols=["validation_id", "model_id", "runtime_secs"]
-    )
-    return results_df
+
+            # Save result immediately
+            new_row_df = pd.DataFrame.from_records([results])
+            results_df = pd.concat([results_df, new_row_df], ignore_index=True)
+            results_df = results_df.drop_duplicates(
+                subset=["validation_id", "model_id"], keep="last"
+            )
+            results_df = results_df.sort_values(by=["validation_id", "model_id"]).reset_index(
+                drop=True
+            )
+            store.write(
+                results_df,
+                results_path,
+                to_front_cols=["validation_id", "model_id", "runtime_secs"],
+            )
+            # Re-index for skip checking on next iteration
+            results_df = results_df.set_index(["validation_id", "model_id"], drop=False)
```

Comment:
can you explain this bit?

Reply:
[c26e989](https://github.com/datavaluepeople/kotsu/pull/66/commits/c26e989bb1a31f287018b80f715e155583e9e1d3)